include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,275 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_AUTO_TUNE_POLICY_H
+#define _ONEDPL_AUTO_TUNE_POLICY_H
+
+#include <atomic>
+#include <exception>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <vector>
+#if _DS_BACKEND_SYCL != 0
+    #include "oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h"
+#endif
+namespace oneapi {
+namespace dpl {
+namespace experimental {
+#if _DS_BACKEND_SYCL != 0
+  template <typename Backend=sycl_backend, typename... KeyArgs>
+#else 
+  template <typename Backend, typename... KeyArgs>
+#endif
+class auto_tune_policy;
+
+  namespace internal { 
+    template <typename Backend, typename Resource, typename Tuner, typename... KeyArgs>
+    class auto_tune_selection_type {
+      using policy_t = auto_tune_policy<Backend, KeyArgs...>;
+      policy_t policy_; 
+      using resource_with_offset_t = Resource;
+      resource_with_offset_t resource_;
+      using tuner_t = Tuner;
+      std::shared_ptr<tuner_t> tuner_;
+
+    public:
+      auto_tune_selection_type() : policy_(deferred_initialization) {} 
+
+      auto_tune_selection_type(const policy_t& p, resource_with_offset_t r, std::shared_ptr<tuner_t> t) 
+        : policy_(p), resource_(r), tuner_(t) {}
+
+      auto unwrap() { return ::oneapi::dpl::experimental::unwrap(resource_.r_); }
+
+      policy_t get_policy() { return policy_; };
+
+      void report(const execution_info::task_time_t&, const typename execution_info::task_time_t::value_type& v) const {
+        tuner_->add_new_timing(resource_, v);
+      }
+    };
+  }
+
+  template <typename Backend, typename... KeyArgs>
+  class auto_tune_policy {
+
+    static constexpr double never_resample = 0.0;
+    static constexpr int use_best_resource = -1;
+
+    using wrapped_resource_t = typename std::decay<Backend>::type::execution_resource_t;
+    using size_type = typename std::vector<typename Backend::resource_type>::size_type;
+
+    using timing_t = uint64_t;
+
+    struct resource_with_offset_t {
+      wrapped_resource_t r_;
+      size_type offset_; 
+    };
+
+    struct time_data_t {
+      uint64_t num_timings_ = 0;
+      timing_t value_ = 0;
+    };
+
+    struct tuner_t {
+      std::mutex m_;
+
+      std::chrono::high_resolution_clock::time_point t0_;
```

Comment:
`high_resolution_clock` can be an alias for either `steady_clock`, `system_clock`, or a different kind of clock. I think it would probably be best to use `steady_clock` in this policy.

Reply:
Ok, we will do that.